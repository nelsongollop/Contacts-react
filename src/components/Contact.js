import React, {useEffect} from 'react'
import { Field, reduxForm } from 'redux-form'
import TextField from "@material-ui/core/TextField";
import Avatar from "@material-ui/core/Avatar";
import Button from '@material-ui/core/Button'
import {connect} from 'react-redux'

const renderTextField = ({input, meta, label}) => (
    <TextField
        {...input}
        id="standard-name"
        label={label}
        margin="normal"
        fullWidth={true}
        helperText={meta.touched ? meta.error : ""}
        FormHelperTextProps={{error: true}}
        onKeyPress={e => {
            if (e.key === 'Enter') e.preventDefault();
        }}
    />
)

let ContactForm = props => {
    const { handleSubmit } = props
    useEffect(() => {
        props.initialize(props.state.selected)
        props.change('id', props.state.selected.id)
    }, [props.state.selectedIndex])

    return (
        <div className="container">
            <div className="row justify-content-md-center">
                <Avatar/>
            </div>
            <div className="row justify-content-md-center">
                <div className="col-md-8">
                    <form onSubmit={handleSubmit}>
                        <Field name="first_name" component={renderTextField} label="First Name"/>
                        <Field name="last_name" component={renderTextField} label="Last Name"/>
                        <Field name="phone" component={renderTextField} label="Phone"/>
                        <Field name="email" component={renderTextField} label="Email"/>

                        <Button className="float-right mt-5" variant="contained" color="secondary" type="submit">
                            {props.state.update ? "Update" : "Create"}
                        </Button>
                    </form>
                </div>
            </div>
        </div>
    )
}

ContactForm = reduxForm({
    form: 'contact'
})(ContactForm)

const mapStateToProps = (state) => {
    return {
        state : state.list
    }
}

export default connect(mapStateToProps)(ContactForm)