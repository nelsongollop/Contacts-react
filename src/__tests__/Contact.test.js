import React from 'react';
import {Provider} from 'react-redux'
import configureStore from 'redux-mock-store';
import {configure, shallow, mount} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
import Contact from "../components/Contact";

configure({ adapter: new Adapter() });
const mockStore = configureStore([]);

describe('Contact component with no selected contact in reducer state', () => {
    let store;
    let wrapper;
    beforeEach(() => {
        store = mockStore({
            list: {
                selectedIndex: null,
                selected: {},
                updating: false
            }
        });
        wrapper = mount(
            <Provider store={store}>
                <Contact />
            </Provider>
        );
    });
    it('should render a contact form', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('The button should say create', () => {
        const button = wrapper.find("button");
        expect(button.text()).toBe("Create");
    });

    it('form should be empty because there is no selected contact', () => {
        const inputs = wrapper.find('input').at(0)
        expect(inputs.text()).toBe('')
    });
});

describe('Contact component with a selected contact in reducer state', () => {
    let wrapper;
    let store;
    beforeEach(() => {
        store = mockStore({
            list: {
                selectedIndex: null,
                selected: {},
                updating: false
            }
        });
        wrapper = mount(
            <Provider store={store}>
                <Contact />
            </Provider>
            );
    });

    it('The field first name from the form should contain the name', () => {
        wrapper.setProps({state: {list: {
                    selectedIndex: 0,
                    selected: {id: '0', first_name: 'Daenerys', last_name: 'Targaryen', phone: '123', email: 'a@a.com'},
                    updating: true
                }}})
        const input = wrapper.find('input[name="first_name"]')
        expect(input.text()).toBe('Daenerys')
    });
});
