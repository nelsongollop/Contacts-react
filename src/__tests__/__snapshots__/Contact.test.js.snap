// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact component with no selected contact in reducer state should render a contact form 1`] = `ReactWrapper {}`;

exports[`Contact component with no selected contact in reducer state should render a list 1`] = `ReactWrapper {}`;

exports[`Contact component with no selected contact in reducer state should render a new contact form 1`] = `ReactWrapper {}`;
